File: clients/go/retry/retry.go
package retry

import (
	"time"

	"github.com/raystack/raccoon/clients/go/log"
)

var (
	// Default retry configuration
	DefaultRetryMax  uint          = 3
	DefaultRetryWait time.Duration = 1 * time.Second
)

// Do retries the function f, waits time between retries until a successful call is made.
// wait is time to wait for the next call
// maxAttempts is the maximum number if calls to the function.
func Do(wait time.Duration, maxAttempts uint, f func() error) error {
	var calls uint = 0
	for {
		err := f()
		if err == nil {
			return nil
		}
		calls++
		if calls >= maxAttempts && maxAttempts != 0 {
			return err
		}
		time.Sleep(wait)
		log.Errorf("[Retry %v] Retrying after %v due to the Error: %v\n", calls, wait, err)
	}
}

File: clients/js/lib/retry/retry.js
async function retry(callback, maxAttempts, waitTime, logger) {
    for (let attempt = 0; attempt < maxAttempts; attempt += 1) {
        try {
            // eslint-disable-next-line no-await-in-loop
            return await callback();
        } catch (error) {
            if (attempt === maxAttempts - 1) {
                throw error;
            }
            // eslint-disable-next-line
            await new Promise((resolve) => setTimeout(resolve, waitTime));
            logger.info(
                `[Retry ${attempt}]: Retrying after ${waitTime} due to the Error: ${error}`
            );
        }
    }
    throw new Error('Retry limit exceeded');
}

export default retry;
